# Answer Key: 60 Practice Questions
## NumPy, Pandas & SQL on `studentsample.csv`

Companion to: `studentsample_practice_questions.md`

Run the setup cell below before working through the answers.

In [3]:
import pandas as pd
import numpy as np
import sqlite3

# Load dataset (Pandas / NumPy sections)
df = pd.read_csv("studentsample.csv")

# SQLite setup (SQL sections F & G)
conn = sqlite3.connect("students.db")
df.to_sql("students", conn, index=False, if_exists="replace")

print("Data loaded:", df.shape)
df.head()

Data loaded: (20, 10)


,student_id,name,age,gender,major,gpa,attendance_pct,city,enrollment_year,scholarship
0,S001,Alice Johnson,18.0,F,Computer Science,3.8,95.0,New York,2022.0,Yes
1,S002,Bob Smith,19.0,M,Mathematics,3.5,88.0,Boston,2021.0,No
2,S003,Charlie Brown,20.0,M,Physics,3.2,92.0,Chicago,2023.0,Yes
3,S004,Diana Prince,21.0,F,Biology,3.9,97.0,Seattle,2022.0,No
4,S005,Ethan Hunt,19.0,M,Chemistry,2.8,85.0,Austin,2021.0,NaN


## Section A: Pandas — Loading & Inspection (1–10)

### 1. Load and display first 5 rows

**Output:** First 5 rows — S001 (Alice Johnson) through S005 (Ethan Hunt).

In [4]:
df = pd.read_csv("studentsample.csv")
df.head()

,student_id,name,age,gender,major,gpa,attendance_pct,city,enrollment_year,scholarship
0,S001,Alice Johnson,18.0,F,Computer Science,3.8,95.0,New York,2022.0,Yes
1,S002,Bob Smith,19.0,M,Mathematics,3.5,88.0,Boston,2021.0,No
2,S003,Charlie Brown,20.0,M,Physics,3.2,92.0,Chicago,2023.0,Yes
3,S004,Diana Prince,21.0,F,Biology,3.9,97.0,Seattle,2022.0,No
4,S005,Ethan Hunt,19.0,M,Chemistry,2.8,85.0,Austin,2021.0,NaN


### 2. Shape of the dataset

**Answer:** `(20, 10)` — 20 rows, 10 columns.

In [5]:
df.shape

(20, 10)

### 3. Column info summary

**Answer:**
- 20 entries, 10 columns
- Numeric: `age`, `gpa`, `attendance_pct`, `enrollment_year` (float64)
- Object/string: `student_id`, `name`, `gender`, `major`, `city`, `scholarship`
- Non-null counts vary (e.g. `scholarship` has 17 non-null, `age` has 18)

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   student_id       20 non-null     str    
 1   name             20 non-null     str    
 2   age              18 non-null     float64
 3   gender           20 non-null     str    
 4   major            18 non-null     str    
 5   gpa              18 non-null     float64
 6   attendance_pct   18 non-null     float64
 7   city             18 non-null     str    
 8   enrollment_year  18 non-null     float64
 9   scholarship      17 non-null     str    
dtypes: float64(4), str(6)
memory usage: 1.7 KB


### 4. Descriptive statistics

**Answer (numeric columns):**

| Stat | age | gpa | attendance_pct | enrollment_year |
|------|-----|-----|----------------|-----------------|
| mean | 19.83 | 3.34 | 88.72 | 2022.0 |
| std  | 1.42  | 0.36 | 5.48  | 0.84 |
| min  | 18.0  | 2.7  | 78.0  | 2021.0 |
| max  | 22.0  | 3.9  | 97.0  | 2023.0 |

In [7]:
df.describe()

,age,gpa,attendance_pct,enrollment_year
count,18.000000,18.000000,18.000000,18.000000
mean,19.833333,3.338889,88.722222,2022.000000
std,1.424574,0.361596,5.475137,0.840168
min,18.000000,2.700000,78.000000,2021.000000
25%,19.000000,3.100000,85.250000,2021.000000
50%,20.000000,3.350000,89.500000,2022.000000
75%,21.000000,3.600000,92.750000,2023.000000
max,22.000000,3.900000,97.000000,2023.000000


### 5. Column names and types

**Answer:**
- **10 columns total**
- **4 numeric:** `age`, `gpa`, `attendance_pct`, `enrollment_year`
- **6 categorical/object:** `student_id`, `name`, `gender`, `major`, `city`, `scholarship`

In [8]:
df.columns.tolist()
df.select_dtypes(include='number').columns   # numeric
df.select_dtypes(exclude='number').columns   # categorical

Index(['student_id', 'name', 'gender', 'major', 'city', 'scholarship'], dtype='str')

### 6. Select specific columns

**Answer:** DataFrame with 20 rows and 3 columns.

In [9]:
df[['name', 'major', 'gpa']]

,name,major,gpa
0,Alice Johnson,Computer Science,3.8
1,Bob Smith,Mathematics,3.5
2,Charlie Brown,Physics,3.2
3,Diana Prince,Biology,3.9
4,Ethan Hunt,Chemistry,2.8
5,Fiona Green,Economics,3.6
6,George Miller,Psychology,NaN
7,Hannah Lee,Engineering,3.4
8,Ian Clark,NaN,3.1
9,Julia Roberts,History,3.7


### 7. Rows 5 through 10 (inclusive)

**Answer:** 6 rows — S006 (Fiona Green) through S011 (Kevin Hart).

In [10]:
df.iloc[5:11]

,student_id,name,age,gender,major,gpa,attendance_pct,city,enrollment_year,scholarship
5,S006,Fiona Green,22.0,F,Economics,3.6,90.0,Denver,2023.0,Yes
6,S007,George Miller,18.0,M,Psychology,NaN,78.0,Miami,2022.0,No
7,S008,Hannah Lee,20.0,F,Engineering,3.4,NaN,Portland,2021.0,Yes
8,S009,Ian Clark,NaN,M,NaN,3.1,91.0,NaN,NaN,No
9,S010,Julia Roberts,21.0,F,History,3.7,86.0,San Diego,2023.0,Yes
10,S011,Kevin Hart,19.0,M,Computer Science,2.9,89.0,New York,2022.0,NaN


### 8. Students with GPA > 3.5

**Answer:** 6 students — S001, S004, S006, S010, S014, S019

| student_id | name | gpa |
|------------|------|-----|
| S001 | Alice Johnson | 3.8 |
| S004 | Diana Prince | 3.9 |
| S006 | Fiona Green | 3.6 |
| S010 | Julia Roberts | 3.7 |
| S014 | Nina Patel | 3.8 |
| S019 | Sara Connor | 3.6 |

In [11]:
df[df['gpa'] > 3.5]

,student_id,name,age,gender,major,gpa,attendance_pct,city,enrollment_year,scholarship
0,S001,Alice Johnson,18.0,F,Computer Science,3.8,95.0,New York,2022.0,Yes
3,S004,Diana Prince,21.0,F,Biology,3.9,97.0,Seattle,2022.0,No
5,S006,Fiona Green,22.0,F,Economics,3.6,90.0,Denver,2023.0,Yes
9,S010,Julia Roberts,21.0,F,History,3.7,86.0,San Diego,2023.0,Yes
13,S014,Nina Patel,18.0,F,Biology,3.8,96.0,Seattle,2022.0,No
18,S019,Sara Connor,22.0,F,Engineering,3.6,90.0,NaN,NaN,Yes


### 9. Female students enrolled in 2023

**Answer:** 3 students — S006 (Fiona Green), S010 (Julia Roberts), S016 (Paula Abdul).

In [12]:
df[(df['gender'] == 'F') & (df['enrollment_year'] == 2023)]

,student_id,name,age,gender,major,gpa,attendance_pct,city,enrollment_year,scholarship
5,S006,Fiona Green,22.0,F,Economics,3.6,90.0,Denver,2023.0,Yes
9,S010,Julia Roberts,21.0,F,History,3.7,86.0,San Diego,2023.0,Yes
15,S016,Paula Abdul,21.0,F,NaN,3.5,93.0,Denver,2023.0,No


### 10. Students per major

**Answer:** Each of the 9 listed majors has **2 students** (2 rows have missing `major`).

In [ ]:
df['major'].value_counts()

## Section B: Pandas — Missing Data (11–20)

*Reload fresh data before Section B:*

In [ ]:
# Reset DataFrame for missing-data exercises
df = pd.read_csv("studentsample.csv")

### 11. Total missing values

**Answer:** `15`

In [ ]:
df.isnull().sum().sum()

### 12. Missing values per column

**Answer:**

| Column | Missing |
|--------|---------|
| student_id, name, gender | 0 |
| age, major, gpa, attendance_pct, city, enrollment_year | 2 each |
| scholarship | **3** (most nulls) |

In [ ]:
df.isnull().sum()

### 13. Rows where age is missing

**Answer:** S009 (Ian Clark), S015 (Oscar Isaac).

In [ ]:
df[df['age'].isnull()]

### 14. Rows with at least one missing value

**Answer:** **11 rows** have at least one null.

In [1]:
df[df.isnull().any(axis=1)]

NameError: name 'df' is not defined

### 15. Rows where both gpa and attendance_pct are missing

**Answer:** **Empty DataFrame** — no student is missing both columns simultaneously.

In [ ]:
df[df['gpa'].isnull() & df['attendance_pct'].isnull()]

### 16. Percentage missing per column

**Answer:** `scholarship` = 15%, all other nullable columns = 10%, `student_id`/`name`/`gender` = 0%.

In [ ]:
(df.isnull().mean() * 100).round(2)

### 17. Drop all rows with any missing value

**Answer:** **9 rows** remain.

In [16]:
df_clean = df.dropna()
len(df_clean)

9

### 18. Drop rows with missing gpa

**Answer:** **18 rows** remain (S007 and S017 removed).

In [15]:
df_step = df.dropna(subset=['student_id', 'name'])  # 0 dropped
df_step = df_step.dropna(subset=['gpa'])
len(df_step)

18

### 19. Fill missing age with median

**Answer:** Median age = **20.0**. S009 and S015 now have `age = 20.0`.

In [ ]:
df['age'] = df['age'].fillna(df['age'].median())

### 20. Fill scholarship and major

**Answer:**
- Mode of `scholarship` = **"Yes"** (fills S005, S011, S020)
- Missing `major` → **"Undeclared"** (S009, S016)

In [ ]:
df['scholarship'] = df['scholarship'].fillna(df['scholarship'].mode()[0])
df['major'] = df['major'].fillna('Undeclared')

## Section C: Pandas — Cleaning & Transformation (21–28)

*Reload fresh data before Section C:*

In [ ]:
# Reset DataFrame for cleaning exercises
df = pd.read_csv("studentsample.csv")

### 21. Convert gpa to float

**Answer:** Column dtype becomes `float64`.

In [ ]:
df['gpa'] = df['gpa'].astype(float)

### 22. Create full_info column

**Answer:** Example: `"Alice Johnson - Computer Science"`

In [ ]:
df['full_info'] = df['name'] + ' - ' + df['major'].fillna('Unknown')

### 23. Extract first name

**Answer:** Example: `"Alice Johnson"` → `"Alice"`

In [ ]:
df['first_name'] = df['name'].str.split().str[0]

### 24. Standardize city

**Answer:** Cities formatted as title case with no extra whitespace.

In [ ]:
df['city'] = df['city'].str.strip().str.title()

### 25. Create gpa_category

**Answer:** High = 8, Medium = 8, Low = 2, Unknown = 2 (S007, S017).

In [ ]:
def categorize_gpa(gpa):
    if pd.isna(gpa):
        return 'Unknown'
    elif gpa >= 3.5:
        return 'High'
    elif gpa >= 3.0:
        return 'Medium'
    else:
        return 'Low'

df['gpa_category'] = df['gpa'].apply(categorize_gpa)

### 26. Rename column

In [ ]:
df = df.rename(columns={'attendance_pct': 'attendance_percent'})

### 27. Check and remove duplicates

**Answer:** **0 duplicate** `student_id` values — no rows removed.

In [ ]:
df.duplicated(subset=['student_id']).sum()   # 0 duplicates
df = df.drop_duplicates(subset=['student_id'])

### 28. Sort and show top 5

**Answer:**

| name | gpa |
|------|-----|
| Diana Prince | 3.9 |
| Alice Johnson | 3.8 |
| Nina Patel | 3.8 |
| Julia Roberts | 3.7 |
| Fiona Green | 3.6 |

In [ ]:
df.sort_values(['gpa', 'name'], ascending=[False, True]).head(5)

## Section D: Pandas — Analysis & Aggregation (29–34)

### 29. Average GPA

**Answer:** **3.34** (approx. 3.3389)

In [ ]:
df['gpa'].mean()

### 30. Average GPA by gender

**Answer:**

| gender | avg gpa |
|--------|---------|
| F | 3.62 |
| M | 3.06 |

In [ ]:
df.groupby('gender')['gpa'].mean()

### 31. Average GPA and attendance by major

**Answer (selected):**

| major | gpa | attendance_pct |
|-------|-----|----------------|
| Biology | 3.85 | 96.5 |
| Economics | 3.60 | 88.5 |
| Chemistry | 2.75 | 82.5 |
| Psychology | 3.20 | 78.0 |

In [ ]:
df.groupby('major')[['gpa', 'attendance_pct']].mean()

### 32. City with highest average GPA

**Answer:** **Seattle** with average GPA **3.85**

In [ ]:
df.groupby('city')['gpa'].mean().idxmax()
df.groupby('city')['gpa'].mean().max()

### 33. Scholarship recipients per enrollment year

**Answer:**

| enrollment_year | count |
|-----------------|-------|
| 2021 | 2 |
| 2022 | 2 |
| 2023 | 4 |

In [ ]:
df[df['scholarship'] == 'Yes'].groupby('enrollment_year').size()

### 34. Pivot table

**Answer:**

| gender | 2021 | 2022 | 2023 |
|--------|------|------|------|
| F | 3.35 | 3.83 | 3.60 |
| M | 3.05 | 2.90 | 3.10 |

In [ ]:
pd.pivot_table(df, values='gpa', index='gender',
               columns='enrollment_year', aggfunc='mean')

## Section E: NumPy — Numerical Operations (35–42)

### 35. GPA array statistics

**Answer:** mean = **3.34**, median = **3.35**, std = **0.35**

In [ ]:
gpa_arr = np.array(df['gpa'], dtype=float)
np.nanmean(gpa_arr), np.nanmedian(gpa_arr), np.nanstd(gpa_arr)

### 36. Min and max GPA

**Answer:** min = **2.7**, max = **3.9**

In [ ]:
np.nanmin(gpa_arr), np.nanmax(gpa_arr)

### 37. Attendance percentiles

**Answer:** 25th = **85.25**, 50th = **89.5**, 75th = **92.75**

In [ ]:
att = df['attendance_pct'].dropna().values
np.percentile(att, [25, 50, 75])

### 38. Age mean and std (ignoring NaN)

**Answer:** mean = **19.83**, std = **1.38**

In [ ]:
np.nanmean(df['age']), np.nanstd(df['age'])

### 39. Replace low GPA with NaN, recalculate mean

**Answer:** New mean = **3.45** (3 values below 3.0 replaced: S005, S011, S015)

In [ ]:
gpa_modified = df['gpa'].copy()
gpa_modified[gpa_modified < 3.0] = np.nan
np.nanmean(gpa_modified)

### 40. Honors array with np.where

**Answer:** `[1,1,0,1,0,1,0,0,0,1,0,0,0,1,0,1,0,0,1,0]` — **8** honors students.

In [ ]:
honors = np.where(df['gpa'] >= 3.5, 1, 0)
honors = np.where(df['gpa'].isna(), 0, honors)

### 41. Correlation between GPA and attendance

**Answer:** Correlation coefficient ≈ **0.72** (positive moderate correlation).

In [ ]:
clean = df[['gpa', 'attendance_pct']].dropna()
np.corrcoef(clean['gpa'], clean['attendance_pct'])

### 42. Normalize GPA to 0–1

**Answer:** Min GPA (2.7) → 0.0, Max GPA (3.9) → 1.0.  
Example: Alice (3.8) → **0.92**, Bob (3.5) → **0.67**

In [13]:
gpa_vals = df['gpa'].dropna()
df.loc[gpa_vals.index, 'gpa_normalized'] = (
    (df.loc[gpa_vals.index, 'gpa'] - gpa_vals.min()) /
    (gpa_vals.max() - gpa_vals.min())
)

## Section F: SQL — Analysis & Cleaning (43–50)

### 43. Students with GPA above 3.5

**Answer:** 6 rows — S001, S004, S006, S010, S014, S019.

**SQL query:**

```sql
SELECT *
FROM students
WHERE gpa > 3.5;
```

In [14]:
print(pd.read_sql("""SELECT *
FROM students
WHERE gpa > 3.5""", conn))

  student_id           name   age gender             major  gpa  \
0       S001  Alice Johnson  18.0      F  Computer Science  3.8   
1       S004   Diana Prince  21.0      F           Biology  3.9   
2       S006    Fiona Green  22.0      F         Economics  3.6   
3       S010  Julia Roberts  21.0      F           History  3.7   
4       S014     Nina Patel  18.0      F           Biology  3.8   
5       S019    Sara Connor  22.0      F       Engineering  3.6   

   attendance_pct       city  enrollment_year scholarship  
0            95.0   New York           2022.0         Yes  
1            97.0    Seattle           2022.0          No  
2            90.0     Denver           2023.0         Yes  
3            86.0  San Diego           2023.0         Yes  
4            96.0    Seattle           2022.0          No  
5            90.0        NaN              NaN         Yes  


### 44. Total student count

**Answer:** `20`

**SQL query:**

```sql
SELECT COUNT(*) AS total_students
FROM students;
```

In [ ]:
print(pd.read_sql("""SELECT COUNT(*) AS total_students
FROM students""", conn))

### 45. Students with missing age

**Answer:** S009 (Ian Clark), S015 (Oscar Isaac).

**SQL query:**

```sql
SELECT *
FROM students
WHERE age IS NULL;
```

In [ ]:
print(pd.read_sql("""SELECT *
FROM students
WHERE age IS NULL""", conn))

### 46. Student count per major

**Answer:** Each major has 2 students; 2 rows have `NULL` major.

**SQL query:**

```sql
SELECT major, COUNT(*) AS student_count
FROM students
GROUP BY major;
```

In [ ]:
print(pd.read_sql("""SELECT major, COUNT(*) AS student_count
FROM students
GROUP BY major""", conn))

### 47. Average GPA and attendance by gender

**Answer:**

| gender | avg_gpa | avg_attendance |
|--------|---------|----------------|
| F | 3.62 | 90.11 |
| M | 3.06 | 87.33 |

**SQL query:**

```sql
SELECT gender,
       AVG(gpa) AS avg_gpa,
       AVG(attendance_pct) AS avg_attendance
FROM students
GROUP BY gender;
```

In [ ]:
print(pd.read_sql("""SE LECT gender,
       AVG(gpa) AS avg_gpa,
       AVG(attendance_pct) AS avg_attendance
FROM students
GROUP BY gender""", conn))

### 48. Scholarship and attendance above 90%

**Answer:** S001 (Alice Johnson, 95%), S003 (Charlie Brown, 92%).

**SQL query:**

```sql
SELECT *
FROM students
WHERE scholarship = 'Yes'
  AND attendance_pct > 90;
```

In [ ]:
print(pd.read_sql("""SELECT *
FROM students
WHERE scholarship = 'Yes'
  AND attendance_pct > 90""", conn))

### 49. Update missing majors

**Answer:** Updates 2 rows (S009, S016).

**SQL:**

```sql
UPDATE students
SET major = 'Undeclared'
WHERE major IS NULL;
```

*Note: UPDATE/DELETE modify the database. Re-run the setup cell to reset `students` table.*

### 50. Delete and top 5 by GPA

**Answer:**
- Part 1: **0 rows deleted** (no student missing both columns).
- Part 2: Diana Prince (3.9), Alice Johnson (3.8), Nina Patel (3.8), Julia Roberts (3.7), Fiona Green (3.6).

**SQL query:**

```sql
-- Part 1: Delete rows where BOTH gpa and attendance_pct are NULL
DELETE FROM students
WHERE gpa IS NULL AND attendance_pct IS NULL;

-- Part 2: Top 5 students by GPA
SELECT *
FROM students
WHERE gpa IS NOT NULL
ORDER BY gpa DESC
LIMIT 5;
```

## Section G: SQL — GROUP BY, HAVING, LIMIT & Stored Procedures (51–60)

### 51. Student count per city

**Answer:** 9 cities, each with **2 students** (tied at top).

**SQL query:**

```sql
SELECT city, COUNT(*) AS student_count
FROM students
WHERE city IS NOT NULL
GROUP BY city
ORDER BY student_count DESC;
```

In [ ]:
print(pd.read_sql("""SELECT city, COUNT(*) AS student_count
FROM students
WHERE city IS NOT NULL
GROUP BY city
ORDER BY student_count DESC""", conn))

### 52. Majors with average GPA > 3.3

**Answer:**

| major | avg_gpa |
|-------|---------|
| Biology | 3.85 |
| Economics | 3.60 |
| Engineering | 3.50 |
| Computer Science | 3.35 |
| History | 3.40 |
| Mathematics | 3.40 |

**SQL query:**

```sql
SELECT major, AVG(gpa) AS avg_gpa
FROM students
WHERE major IS NOT NULL
GROUP BY major
HAVING AVG(gpa) > 3.3;
```

In [ ]:
print(pd.read_sql("""SELECT major, AVG(gpa) AS avg_gpa
FROM students
WHERE major IS NOT NULL
GROUP BY major
HAVING AVG(gpa) > 3.3""", conn))

### 53. Enrollment years with 3+ students

**Answer:**

| enrollment_year | student_count |
|-----------------|---------------|
| 2021 | 6 |
| 2022 | 6 |
| 2023 | 6 |

**SQL query:**

```sql
SELECT enrollment_year, COUNT(*) AS student_count
FROM students
WHERE enrollment_year IS NOT NULL
GROUP BY enrollment_year
HAVING COUNT(*) >= 3;
```

In [ ]:
print(pd.read_sql("""SELECT enrollment_year, COUNT(*) AS student_count
FROM students
WHERE enrollment_year IS NOT NULL
GROUP BY enrollment_year
HAVING COUNT(*) >= 3""", conn))

### 54. Top 3 by attendance

**Answer:**

| name | attendance_pct |
|------|----------------|
| Diana Prince | 97 |
| Nina Patel | 96 |
| Alice Johnson | 95 |

**SQL query:**

```sql
SELECT name, attendance_pct
FROM students
WHERE attendance_pct IS NOT NULL
ORDER BY attendance_pct DESC
LIMIT 3;
```

In [ ]:
print(pd.read_sql("""SELECT name, attendance_pct
FROM students
WHERE attendance_pct IS NOT NULL
ORDER BY attendance_pct DESC
LIMIT 3""", conn))

### 55. Top 2 majors by average GPA

**Answer:**

| major | avg_gpa |
|-------|---------|
| Biology | 3.85 |
| Economics | 3.60 |

**SQL query:**

```sql
SELECT major, AVG(gpa) AS avg_gpa
FROM students
WHERE major IS NOT NULL AND gpa IS NOT NULL
GROUP BY major
ORDER BY avg_gpa DESC
LIMIT 2;
```

In [ ]:
print(pd.read_sql("""SELECT major, AVG(gpa) AS avg_gpa
FROM students
WHERE major IS NOT NULL AND gpa IS NOT NULL
GROUP BY major
ORDER BY avg_gpa DESC
LIMIT 2""", conn))

### 56. Cities with more than 1 student

**Answer:** All 9 cities with data (each has exactly 2 students).

**SQL query:**

```sql
SELECT city, COUNT(*) AS student_count
FROM students
WHERE city IS NOT NULL
GROUP BY city
HAVING COUNT(*) > 1;
```

In [ ]:
print(pd.read_sql("""SELECT city, COUNT(*) AS student_count
FROM students
WHERE city IS NOT NULL
GROUP BY city
HAVING COUNT(*) > 1""", conn))

### 57. Count by gender and enrollment year

**Answer:**

| gender | enrollment_year | count |
|--------|-----------------|-------|
| F | 2021 | 2 |
| M | 2021 | 4 |
| F | 2022 | 4 |
| M | 2022 | 2 |
| F | 2023 | 3 |
| M | 2023 | 3 |

**SQL query:**

```sql
SELECT gender, enrollment_year, COUNT(*) AS student_count
FROM students
WHERE enrollment_year IS NOT NULL
GROUP BY gender, enrollment_year
ORDER BY enrollment_year, gender;
```

In [ ]:
print(pd.read_sql("""SELECT gender, enrollment_year, COUNT(*) AS student_count
FROM students
WHERE enrollment_year IS NOT NULL
GROUP BY gender, enrollment_year
ORDER BY enrollment_year, gender""", conn))

### 58. Majors with max GPA >= 3.8

**Answer:**

| major | max_gpa |
|-------|---------|
| Biology | 3.9 |
| Computer Science | 3.8 |

**SQL query:**

```sql
SELECT major, MAX(gpa) AS max_gpa
FROM students
WHERE major IS NOT NULL
GROUP BY major
HAVING MAX(gpa) >= 3.8;
```

In [ ]:
print(pd.read_sql("""SELECT major, MAX(gpa) AS max_gpa
FROM students
WHERE major IS NOT NULL
GROUP BY major
HAVING MAX(gpa) >= 3.8""", conn))

### 59. Stored procedure — GetStudentsByMajor (MySQL)

**Answer:** Returns S001 (Alice Johnson) and S011 (Kevin Hart).

**SQL (MySQL / PostgreSQL):**

```sql
DELIMITER //

CREATE PROCEDURE GetStudentsByMajor(IN p_major VARCHAR(100))
BEGIN
    SELECT *
    FROM students
    WHERE major = p_major;
END //

DELIMITER ;

-- Call:
CALL GetStudentsByMajor('Computer Science');
```

### 60. Stored procedure — CountLowGpaStudents (MySQL)

**Answer:** Returns **3** (S005 gpa=2.8, S011 gpa=2.9, S015 gpa=2.7).

**SQL (MySQL / PostgreSQL):**

```sql
DELIMITER //

CREATE PROCEDURE CountLowGpaStudents(IN p_gpa DECIMAL(3,2))
BEGIN
    SELECT COUNT(*) AS low_gpa_count
    FROM students
    WHERE gpa IS NOT NULL
      AND gpa < p_gpa;
END //

DELIMITER ;

-- Call:
CALL CountLowGpaStudents(3.0);
```

### Bonus: PostgreSQL function (Question 59)

**SQL (MySQL / PostgreSQL):**

```sql
CREATE OR REPLACE FUNCTION get_students_by_major(p_major TEXT)
RETURNS TABLE (
    student_id TEXT,
    name TEXT,
    age DOUBLE PRECISION,
    gender TEXT,
    major TEXT,
    gpa DOUBLE PRECISION,
    attendance_pct DOUBLE PRECISION,
    city TEXT,
    enrollment_year DOUBLE PRECISION,
    scholarship TEXT
) AS $$
BEGIN
    RETURN QUERY
    SELECT s.student_id, s.name, s.age, s.gender, s.major,
           s.gpa, s.attendance_pct, s.city, s.enrollment_year, s.scholarship
    FROM students s
    WHERE s.major = p_major;
END;
$$ LANGUAGE plpgsql;

-- Call:
SELECT * FROM get_students_by_major('Computer Science');
```